pkg/service/gcp/node.go

```diff
@@ -0,0 +1,166 @@
+package gcp
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"gitlab.com/netbook-devs/spawner-service/pkg/service/common"
+	"gitlab.com/netbook-devs/spawner-service/pkg/service/constants"
+	"gitlab.com/netbook-devs/spawner-service/pkg/service/labels"
+	"gitlab.com/netbook-devs/spawner-service/proto/netbookai/spawner"
+	proto "gitlab.com/netbook-devs/spawner-service/proto/netbookai/spawner"
+	container_proto "google.golang.org/genproto/googleapis/container/v1"
+)
+
+//getNodeFQN retrive fqdn for the node
+func getNodePoolFQN(projectId, location, clusterId, nodeId string) string {
+	return fmt.Sprintf("projects/%s/locations/%s/clusters/%s/nodePools/%s", projectId, location, clusterId, nodeId)
+}
+
+func getDiskType() string {
+	// Doc : https://cloud.google.com/compute/docs/disks
+	return "pd-standard"
+}
+
+//getNodePool Get the NodePool config for the given NodeSpec
+func getNodePool(node *spawner.NodeSpec) (*container_proto.NodePool, error) {
+
+	nodeCount := int32(1)
+	if node.Count != 0 {
+		nodeCount = int32(node.Count)
+	}
+	//node labels
+	label := make(map[string]string)
+	for k, v := range labels.GetNodeLabel(node) {
+		label[k] = *v
+	}
+
+	instance := ""
+	if node.MachineType != "" {
+		instance = common.GetInstance(constants.GcpLabel, node.MachineType)
+	} else {
+		instance = node.Instance
+	}
+
+	if instance == "" {
+		return nil, errors.New(constants.InvalidInstanceOrMachineType)
+	}
+	//Doc : https://cloud.google.com/kubernetes-engine/docs/concepts/node-images#available_node_images
+	imageType := "COS_CONTAINERD"
+	diskType := getDiskType()
+
+	nodeConfig := &container_proto.NodeConfig{
+
+		MachineType: instance,
+		DiskSizeGb:  node.DiskSize,
+		ImageType:   imageType,
+		Preemptible: false,
+		DiskType:    diskType,
+		Labels:      label,
+	}
+	np := container_proto.NodePool{
+
+		Name:             node.Name,
+		Config:           nodeConfig,
+		InitialNodeCount: nodeCount,
+		//in case we use Zonal, if not specied gcp will create per zone node, so 1/2, resulting in 3 nodes default,
+		//us-central1-a, us-central1-b, us-central1-c, us-central1-f
+		//FIXME: should get the node zones from the regions
+		Locations: []string{"us-central1-a"},
```

Comment:
Instead of hardcoding to a specific zone can we do something like `fmt.Sprintf("%s-a", node.region`? Assuming all regions have at least 1 zone and it is `-a`

Reply:
open issue